File: Typecast/TCFontCollection.m
//
//  TCFontCollection.m
//  Typecast
//
//  Created by David Schweinsberg on 1/12/12.
//  Copyright (c) 2012 David Schweinsberg. All rights reserved.
//

#import "TCFontCollection.h"
#import "TCDataInput.h"
#import "TCResourceHeader.h"
#import "TCResourceMap.h"
#import "TCResourceType.h"
#import "TCResourceReference.h"
#import "TCFont.h"
#import "TCTTCHeader.h"

@implementation TCFontCollection

- (id)initWithData:(NSData *)data isSuitcase:(BOOL)suitcase
{
    self = [super init];
    if (self)
    {
        _suitcase = suitcase;
        [self loadFromData:data];
        if (_fonts == nil)
            return nil;
    }
    return self;
}

- (void)loadFromData:(NSData *)fontData
{
    TCDataInput *dataInput = [[TCDataInput alloc] initWithData:fontData];

    if ([self isSuitcase])
    {
        // This is a Macintosh font suitcase resource
        TCResourceHeader *resourceHeader = [[TCResourceHeader alloc] initWithDataInput:dataInput];

        // Seek to the map offset and read the map
        [dataInput reset];
        [dataInput skipByteCount:[resourceHeader mapOffset]];
        TCResourceMap *map = [[TCResourceMap alloc] initWithDataInput:dataInput];

        // Get the 'sfnt' resources
        TCResourceType *resourceType = [map resourceTypeWithName:@"sfnt"];

        // This could be a font suitcase, but with only FONT or NFNT resources
        if (resourceType == nil)
            return;

        // Load the font data
        NSMutableArray *fonts = [[NSMutableArray alloc] initWithCapacity:[resourceType count]];
        for (TCResourceReference *resourceReference in [resourceType references])
        {
            TCFont *font = [[TCFont alloc] init];
            [fonts addObject:font];
            int offset = [resourceHeader dataOffset] + [resourceReference dataOffset] + 4;
            [font readFromDataInput:dataInput directoryOffset:offset tablesOrigin:offset];
        }
        _fonts = fonts;

    }
    else if ([TCTTCHeader isTTCDataInput:dataInput])
    {
        // This is a TrueType font collection
        [dataInput reset];
        _ttcHeader = [[TCTTCHeader alloc] initWithDataInput:dataInput];
        NSMutableArray *fonts = [[NSMutableArray alloc] initWithCapacity:[_ttcHeader directoryCount]];
        for (int i = 0; i < [_ttcHeader directoryCount]; ++i)
        {
            TCFont *font = [[TCFont alloc] init];
            [fonts addObject:font];
            int offset = [[_ttcHeader tableDirectory][i] intValue];
            [font readFromDataInput:dataInput directoryOffset:offset tablesOrigin:0];
        }
        _fonts = fonts;
    }
    else
    {
        // This is a standalone font file
        TCFont *font = [[TCFont alloc] init];
        [font readFromDataInput:dataInput directoryOffset:0 tablesOrigin:0];
        _fonts = @[font];
    }
}

@end

File: Typecast/TCFontCollection.swift
//
//  TCFontCollection.swift
//  Type Designer
//
//  Created by David Schweinsberg on 7/27/17.
//  Copyright © 2017 David Schweinsberg. All rights reserved.
//

import Foundation

class TCFontCollection: NSObject {
  var fonts: [TCFont]
  var ttcHeader: TCTTCHeader?
  var suitcase: Bool

  init(data: Data, isSuitcase: Bool) {
    fonts = []
    self.suitcase = isSuitcase
    super.init()

    if suitcase {
      // This is a Macintosh font suitcase resource
      let dataInput = TCDataInput(data: data)
      let resourceHeader = TCResourceHeader(dataInput: dataInput)

      // Seek to the map offset and read the map
      let mapData = data.subdata(in: Int(resourceHeader.mapOffset)..<data.count)
      let mapDataInput = TCDataInput(data: mapData)
      let map = TCResourceMap(dataInput: mapDataInput)

      // Get the 'sfnt' resources
      let resourceType = map.resourceType(name: "sfnt")

      // This could be a font suitcase, but with only FONT or NFNT resources
      if resourceType == nil {
        return
      }

      // Load the font data
      for resourceReference in (resourceType?.references)! {
        let offset = resourceHeader.dataOffset + UInt32(resourceReference.dataOffset + 4)
        let resData = data.subdata(in: Int(offset)..<data.count)
        if let font = TCFont(data: resData, tablesOrigin: 0) {
          fonts.append(font)
        }
      }
    } else if TCTTCHeader.isTTC(data: data) {

      // This is a TrueType font collection
      ttcHeader = TCTTCHeader(data: data)
      for i in 0 ..< Int((ttcHeader?.directoryCount)!) {
        let offset = ttcHeader?.tableDirectory[i]
//        font.read(dataInput: dataInput, directoryOffset: Int(offset), tablesOrigin: 0)
        let fontData = data.subdata(in: Int(offset!)..<data.count)
        if let font = TCFont(data: fontData, tablesOrigin: UInt(offset!)) {
          fonts.append(font)
        }
      }
    } else {

      // This is a standalone font file
      if let font = TCFont(data: data, tablesOrigin: 0) {
        fonts.append(font)
      }
    }
  }
}
